drivers/counter/counter_mcux_qtmr.c

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright (c) 2022 KT-Elektronik, Klaucke und Partner GmbH
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * Counter driver for the Quad Timer through the MCUxpresso SDK. Based mainly on counter_mcux_gpt.c
+ *
+ * Each quad timer module has four channels (0-3) that can operate independently, but the Zephyr
+ * counter-API does not support starting or stopping different channels independently. Hence, each
+ * channel is represented as an independent counter device.
+ */
+
+#include <drivers/counter.h>
+#include <drivers/clock_control.h>
+#include <fsl_qtmr.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(mcux_qtmr, CONFIG_COUNTER_LOG_LEVEL);
+
+struct mcux_qtmr_config {
+	/* info must be first element */
+	struct counter_config_info info;
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	TMR_Type *base;
+	clock_name_t clock_source;
+	qtmr_channel_selection_t channel;
+	qtmr_config_t qtmr_config;
+	qtmr_counting_mode_t mode;
+};
+
+struct mcux_qtmr_data {
+	counter_alarm_callback_t alarm_callback;
+	counter_top_callback_t top_callback;
+	void *alarm_user_data;
+	void *top_user_data;
+	qtmr_status_flags_t interrupt_mask;
+	uint32_t freq;
+};
+
+/* Only one interrupt per QTMR module. Each of which has four timers. */
+#define DT_DRV_COMPAT nxp_imx_qtmr
+
+/**
+ * @brief ISR for a specific timer channel
+ *
+ * @param dev	timer channel device
+ */
+void mcux_qtmr_timer_handler(const struct device *dev, uint32_t status)
+{
+	const struct mcux_qtmr_config *config = dev->config;
+	struct mcux_qtmr_data *data = dev->data;
+	uint32_t current = QTMR_GetCurrentTimerCount(config->base, config->channel);
+
+	QTMR_ClearStatusFlags(config->base, config->channel, status);
+	__DSB();
+
+	if ((status & kQTMR_Compare1Flag) && data->alarm_callback) {
+		QTMR_DisableInterrupts(config->base, config->channel,
+			kQTMR_Compare1InterruptEnable);
+		data->interrupt_mask &= ~kQTMR_Compare1InterruptEnable;
+		counter_alarm_callback_t alarm_cb = data->alarm_callback;
+
+		data->alarm_callback = NULL;
+		alarm_cb(dev, config->channel, current, data->alarm_user_data);
+	}
+
+	if ((status & kQTMR_OverflowFlag) && data->top_callback) {
+		data->top_callback(dev, data->top_user_data);
+	}
+}
+
+/**
+ * @brief ISR for the QTMR
+ *
+ * @param timers	array containing the counter devices for each channel of the timer module
+ */
+static void mcux_qtmr_isr(const struct device *timers[])
+{
+	/* the interrupt can be triggered by any of the four channels of the QTMR. Check status
+	 * of all channels and trigger the ISR for the channel(s) that has/have triggered the
+	 * interrupt.
+	 */
+	for (qtmr_channel_selection_t ch = kQTMR_Channel_0; ch <= kQTMR_Channel_3 ; ch++) {
+		if (timers[ch] != NULL) {
+			const struct mcux_qtmr_config *config = timers[ch]->config;
+			struct mcux_qtmr_data *data = timers[ch]->data;
+
+			uint32_t channel_status = QTMR_GetStatus(config->base, ch);
```

Comment:
Seems unnecessary to have this in the for loop as you need to get the status reg only once.

Reply:
actually, I have to get the status once for every channel (note that the loop variable `ch` is an argument to this function call).

Reply:
Thank you. I missed that there is a status register per channel